Add docs for using notation-azure-kv (#7)
Azure/notation-azure-kv: docs/nv2-sign-verify-aks.md

# Notary v2 - Remote Signing and Verification with Gatekeeper, Ratify and AKS

> ## Install the notation cli and akv plugin

Reviewer: ```suggestion
## Install the notation cli and azure-kv plugin
```

> NOTE: These are temporary steps for early development. Officially released binaries will be available as the latest changes are merged.

1. Build a version of notation that supports extensibility

    > Note: See [feat: add initial framework for notation extensibility #142](https://github.com/notaryproject/notation/pull/142#issuecomment-996356691) for this PR merging into [notation/tree/feat-kv-extensibility](https://github.com/notaryproject/notation/tree/feat-kv-extensibility)
    ```bash
    git clone git@github.com:aramase/notation.git -b plugin
    cd notation
    make build
    # Copy the notation cli to your bin directory
    cp ./bin/notation ~/bin
    ```

> 2. Build the notation-akv plugin for remote signing and verification

Reviewer: ```suggestion
2. Build the notation-azure-kv plugin for remote signing and verification
```

    git clone git@github.com:Azure/notation-azure-kv.git
    cd notation-azure-kv
    make build

    # Copy the plugin to the notation plugin directory
    mkdir -p ~/.config/notation/plugins/azure-kv
    cp ./bin/notation-azure-kv ~/.config/notation/plugins/azure-kv
     ```

## Configure Environment Variables

To ease the execution of the commands to complete this article, provide values for the Azure resources.

1. Configure Azure Resource Names and Locations

    ```bash
    # location resources will be created in. During Preview of ORAS Artifact support, SouthCentralUS is the only supported region.
    LOCATION=southcentralus

    # Name of the registry example: myregistry.azurecr.io
    ACR_NAME=myregistry
    # Name of the ACR Resource Group
    ACR_RG=${ACR_NAME}-acr-rg
    # Full domain of the ACR
    REGISTRY=$ACR_NAME.azurecr.io
    
    # Name of the Azure Key Vault used to store the signing keys
    AKV_NAME=myakv
    # Key name used to sign and verify
    KEY_NAME=wabbit-networks-io
    KEY_SUBJECT_NAME=wabbit-networks.io
    # Name of the AKV Resource Group
    AKV_RG=${AKV_NAME}-akv-rg

    # Name of the Azure Kubernetes Service instance
    AKS_NAME=myaks

>     AKS_RG=${AKS_NAME}--aks-rg

Reviewer: ```suggestion
    AKS_RG=${AKS_NAME}-aks-rg
```

    ```azurecli
    az group create --name $AKV_RG --location $LOCATION
    az keyvault create --name $AKV_NAME --resource-group $AKV_RG --location $LOCATION
    ```

2. Create a service principal

    ```bash
    # Service Principal Name
    SP_NAME=https://${AKV_NAME}-sp

    # Create the service principal, capturing the password
    export AZURE_CLIENT_SECRET=$(az ad sp create-for-rbac --skip-assignment --name $SP_NAME --query "password" --output tsv)

    # Capture the service srincipal appId
    export AZURE_CLIENT_ID=$(az ad sp list --display-name $SP_NAME --query "[].appId" --output tsv)

    # Capture the Azure Tenant ID
    export AZURE_TENANT_ID=$(az account show --query "tenantId" -o tsv)
    ```

3. Assign key and certificate permissions to the service principal object id

    ```azure-cli
    az keyvault set-policy --name $AKV_NAME --key-permissions get sign --spn $AZURE_CLIENT_ID

    az keyvault set-policy --name $AKV_NAME --certificate-permissions get --spn $AZURE_CLIENT_ID
    ```

## Configure the notation Azure Key Vault plugin

1. Add the Plugin

    ```bash

>     notation plugin add azure-kv ~/.config/notation/plugins/azure-kv/notation-akv

Reviewer: ```suggestion
    notation plugin add azure-kv ~/.config/notation/plugins/azure-kv/notation-azure-kv
```

    ```azurecli-interactive
    az aks create -n $AKS_NAME -g $AKS_RG #--attach-acr $ACR_NAME
    az aks update -n $AKS_NAME -g $AKS_RG --attach-acr $ACR_NAME
    ```

3. Get the AKS credentials

    ```azurecli-interactive
    az aks get-credentials -n $AKS_NAME -g $AKS_RG
    ```

3. Create an ACR Token for notation signing and the ORAS cli to access the registry

    ```azure-cli
    export NOTATION_USERNAME=$ACR_NAME'-token'
    export NOTATION_PASSWORD=$(az acr token create -n $NOTATION_USERNAME \
                        -r $ACR_NAME \
                        --scope-map _repositories_admin \
                        --only-show-errors \
                        -o json | jq -r ".credentials.passwords[0].value")
    ```

4. Configure permissions for ratify
    > NOTE: these are temporary steps. Ratify should use the node secrets, using ACR
    ```bash
    kubectl create secret docker-registry regcred \
        --docker-server=$REGISTRY \
        --docker-username=$NOTATION_USERNAME \
        --docker-password=$NOTATION_PASSWORD \
        --docker-email=someone@example.com
    ```

## Install Gatekeeper

In this step, Gatekeeper will be configured, enabling deployment policies.

> 1. install Gatekeeper

Reviewer: ```suggestion
1. Install Gatekeeper
```

    ```azurecli-interactive
    helm repo add gatekeeper https://open-policy-agent.github.io/gatekeeper/charts

    helm install gatekeeper/gatekeeper  \
        --name-template=gatekeeper \
        --namespace gatekeeper-system --create-namespace \
        --set enableExternalData=true \
        --set controllerManager.dnsPolicy=ClusterFirst,audit.dnsPolicy=ClusterFirst
    ````
 
## Build, Validate, Deploy

We're starting with the following basic Azure service configurations

- An ACR with ORAS Artifacts Support, enabling graphs of supply chain artifacts
- An AKS instance, linked to ACR for pulling images
- AKS with Gatekeeper support

We'll now demo:

1. Creating a cert in Azure Key Vault, used for remote signing

>    > NOTE: The cert can be a CA issues, or self-signed cert. It's really up to the the user whether they need revocation for their controlled environment scenarios

Reviewer: ```suggestion
   > NOTE: The cert can be a CA issued, or self-signed cert. It's really up to the the user whether they need revocation for their controlled environment scenarios
```

2. Secure a namespace with images with trusted signatures
3. Build an image
4. Remotely sign the image with Azure Key Vault
5. Deploy the image to AKS - with a success
6. Attempt to deploy an unsigned image, and watch it fail.

> ### Deploy an image

Reviewer: Should this be `## Deploy an image`? Header 2 instead of Header 3?

Author: It was intended to be nested under the Build, Validate, Deploy heading.

> 1. Create a namespace in AKS

Reviewer: ```suggestion
1. Create a namespace in AKS cluster
```

>     ```azurecli-interactive

Reviewer: ```suggestion
    ```bash
```

    kubectl create ns freezone
    ```

2. Deploy `hello-world:latest` to show a functioning cluster

    ```bash
    kubectl run hello-world --image=mcr.microsoft.com/azuredocs/aci-helloworld:latest -n freezone
    ```

3. List the running pods

    ```bash

>     kubectl get pods -A

Reviewer: ```suggestion
    kubectl get pods -n freezone
```

Since this section is only for showing the pod running successfully in `freezone` namespace, it might be good to just list pods in that namespace and validate.
